go-controller/pkg/allocator/udn/allocator.go

```diff
@@ -20,6 +23,40 @@ func (a *Allocator) AllocateVRFTable(networkID int) (uint, error) {
 	return a.calculateNetworkIDs("vrf-table", networkID, config.UserDefinedNetworks.VRFTableBase, config.UserDefinedNetworks.MaxNetworks, 1)
 }
 
+func (a *Allocator) AllocateV4MasqueradeIPs(networkID int) ([]net.IP, error) {
+	return a.allocateMasqueradeIPs(networkID, "v4", config.Gateway.V4MasqueradeSubnet)
+}
+
+func (a *Allocator) AllocateV6MasqueradeIPs(networkID int) ([]net.IP, error) {
+	return a.allocateMasqueradeIPs(networkID, "v6", config.Gateway.V6MasqueradeSubnet)
+}
+
+func (a *Allocator) allocateMasqueradeIPs(networkID int, idPrefix, masqueradeSubnet string) ([]net.IP, error) {
+	numberOfIPs := uint(2)
+	firstID, err := a.calculateNetworkIDs(idPrefix+"-masquerade-subnets", networkID, config.UserDefinedNetworks.MasqueradeIPBase, config.UserDefinedNetworks.MaxNetworks*numberOfIPs, numberOfIPs)
+	if err != nil {
+		return nil, err
+	}
+	ip, netCidr, err := net.ParseCIDR(masqueradeSubnet)
+	if err != nil {
+		return nil, err
+	}
+
+	masqueradeIPs := []net.IP{}
+	baseIPBigInt := utilnet.BigForIP(ip)
+	for i := uint(0); i < numberOfIPs; i++ {
+		nextIP := utilnet.AddIPOffset(baseIPBigInt, int(firstID+i))
```

Comment:
> this seems fragile to me. It looks like the previous calculateNetworkIDs function is ensuring that there ill be a big enough gap between numbers that there wont be overlap, derived from the ID passed to it and cardinality. Then you use that to know which IPs you can use. I think the problem with this strategy is if callers dont use the same cardinality, wont it overlap?
> 
> like if I say base 10, id 1, cardinality 2: 10+1 + 2*1-2= 11, and now i think i can allocate 11 and 12.
> 
> Now someone else with ID 2 calls the function with base 10, id 2, cardinality 1: 10+1+1*2-1=12
> 
> Now there is overlap. Is my understanding correct? If so, and this is the direction we want to go in we need to put a lot more comments around the caluclateNetworkIDs function.
> 

That would be like having ct mark  12  and masquerade ip 12,13 -> and that's perfectly fine, the cardinality is an argument of a private method, users may add more public methods for different aspects of UDN and they can collide since they are different concept, ID should not collide within their domain, ej ct-mark or masquerade IP, table ID, etc...

> Additionally, need to think about what the cost is of not using a real allocator is. For example, today we allocate IPs and then store them as annotations on kubernetes objects. On OVNK start up, we preallocate all of those IPs to ensure they are not handed out to new pods or nodes. I'm not sure what happens in that case if the IP is already allocated by a node/pod, and another node/pod has a conflicting IP address. We should look into it and see what happens.
> 

From whave we have learn with the `active-network` namespace annotation, using annotations to is a security problem and we have already a number that is unique per network, the `network-id` generated by the bitmap allocator, so we just need to use that number to derive a unique masquerade IP.

About colliding at nodes same can happend today by users configuring networking at nodes with current masquerades, I understand that we should document what users should not do to not break ovn-k.

> If we use this allocator to allocate masquerade IPs, then I'm not sure if we will annotate them to the network object or if it is not necessary. However, if we are annotating join subnet or something like that, and someone modified the annotation or there was a collision would a real allocator help us in anyway that this method does not?

To do that we would need to update that real allocator constantly from the node network configuration to be sure that users has no configure something there that collidas with our topology.

Use documentation about what things should not be done at nodes/pods to not break ovn-k should be enough.